File: smoketests/tests/rls.py
import logging

from .. import Smoketest, random_string

class Rls(Smoketest):
    MODULE_CODE = """
use spacetimedb::{Identity, ReducerContext, Table};

#[spacetimedb::table(name = users, public)]
pub struct Users {
    name: String,
    identity: Identity,
}

#[spacetimedb::client_visibility_filter]
const USER_FILTER: spacetimedb::Filter = spacetimedb::Filter::Sql(
    "SELECT * FROM users WHERE identity = :sender"
);

#[spacetimedb::reducer]
pub fn add_user(ctx: &ReducerContext, name: String) {
    ctx.db.users().insert(Users { name, identity: ctx.sender() });
}
"""

    def test_rls_rules(self):
        """Tests for querying tables with RLS rules"""

        # Insert an identity for Alice
        self.call("add_user", "Alice")

        # Insert a new identity for Bob
        self.reset_config()
        self.new_identity()
        self.call("add_user", "Bob")

        # Query the users table using Bob's identity
        self.assertSql("SELECT name FROM users", """\
 name
-------
 "Bob"
""")
        
        # Query the users table using a new identity
        self.reset_config()
        self.new_identity()
        self.assertSql("SELECT name FROM users", """\
 name
------
""")

class BrokenRls(Smoketest):
    AUTOPUBLISH = False

    MODULE_CODE_BROKEN = """
use spacetimedb::{client_visibility_filter, Filter};

#[spacetimedb::table(name = user)]
pub struct User {
    identity: Identity,
}

#[client_visibility_filter]
const PERSON_FILTER: Filter = Filter::Sql("SELECT * FROM \"user\" WHERE identity = :sender");
"""

    def test_publish_fails_for_rls_on_private_table(self):
        """This tests that publishing an RLS rule on a private table fails"""

        name = random_string()

        self.write_module_code(self.MODULE_CODE_BROKEN)

        with self.assertRaises(Exception):
            self.publish_module(name)

class DisconnectRls(Smoketest):
    AUTOPUBLISH = False
    
    MODULE_CODE = """
use spacetimedb::{Identity, ReducerContext, Table};

#[spacetimedb::table(name = users, public)]
pub struct Users {
    name: String,
    identity: Identity,
}

#[spacetimedb::reducer]
pub fn add_user(ctx: &ReducerContext, name: String) {
    ctx.db.users().insert(Users { name, identity: ctx.sender() });
}
"""
    
    ADD_RLS = """ 
#[spacetimedb::client_visibility_filter]
const USER_FILTER: spacetimedb::Filter = spacetimedb::Filter::Sql(
    "SELECT * FROM users WHERE identity = :sender"
);
"""
    
    def assertSql(self, sql, expected):
        self.maxDiff = None
        sql_out = self.spacetime("sql", self.database_identity, sql)
        sql_out = "\n".join([line.rstrip() for line in sql_out.splitlines()])
        expected = "\n".join([line.rstrip() for line in expected.splitlines()])
        self.assertMultiLineEqual(sql_out, expected)
    
    def test_rls_disconnect_if_change(self):
        """This tests that changing the RLS rules disconnects existing clients"""
        
        name = random_string()
        
        self.write_module_code(self.MODULE_CODE)
        
        self.publish_module(name)
        logging.info("Initial publish complete")
        
        # Now add the RLS rules
        self.write_module_code(self.MODULE_CODE + self.ADD_RLS)
        self.publish_module(name, clear=False, break_clients=True)
        
        # Check the row-level SQL filter is added correctly
        self.assertSql(
            "SELECT sql FROM st_row_level_security",
            """\
 sql
------------------------------------------------
 "SELECT * FROM users WHERE identity = :sender"
""",
        )
        
        logging.info("Re-publish with RLS complete")
        
        logs = self.logs(100)
        
        # Validate disconnect + schema migration logs
        self.assertIn("Disconnecting all users", logs)
    
    def test_rls_no_disconnect(self):
        """This tests that not changing the RLS rules does not disconnect existing clients"""
        
        name = random_string()
        
        self.write_module_code(self.MODULE_CODE + self.ADD_RLS)
        
        self.publish_module(name)
        logging.info("Initial publish complete")
        
        # Now re-publish the same module code
        self.publish_module(name, clear=False, break_clients=False)
        
        logging.info("Re-publish without RLS change complete")
        
        logs = self.logs(100)
        
        # Validate no disconnect logs
        self.assertNotIn("Disconnecting all users", logs)

File: crates/smoketests/tests/rls.rs
use spacetimedb_smoketests::Smoketest;

/// Tests for querying tables with RLS rules
#[test]
fn test_rls_rules() {
    let test = Smoketest::builder().precompiled_module("rls").build();

    // Insert a user for Alice (current identity)
    test.call("add_user", &["Alice"]).unwrap();

    // Create a new identity for Bob
    test.new_identity().unwrap();
    test.call("add_user", &["Bob"]).unwrap();

    // Query the users table using Bob's identity - should only see Bob
    test.assert_sql(
        "SELECT name FROM users",
        r#" name
-------
 "Bob""#,
    );

    // Create another new identity - should see no users
    test.new_identity().unwrap();
    test.assert_sql(
        "SELECT name FROM users",
        r#" name
------"#,
    );
}

/// Module code with RLS on a private table (intentionally broken)
const MODULE_CODE_BROKEN_RLS: &str = r#"
use spacetimedb::{client_visibility_filter, Filter, Identity};

#[spacetimedb::table(accessor = user)]
pub struct User {
    identity: Identity,
}

#[client_visibility_filter]
const PERSON_FILTER: Filter = Filter::Sql("SELECT * FROM \"user\" WHERE identity = :sender");
"#;

/// Tests that publishing an RLS rule on a private table fails
#[test]
fn test_publish_fails_for_rls_on_private_table() {
    let mut test = Smoketest::builder()
        .module_code(MODULE_CODE_BROKEN_RLS)
        .autopublish(false)
        .build();

    let name = format!("test-db-{}", std::process::id());

    // Publishing should fail because RLS is on a private table
    let result = test.publish_module_named(&name, false);
    assert!(result.is_err(), "Expected publish to fail for RLS on private table");
}

/// Tests that changing the RLS rules disconnects existing clients
#[test]
fn test_rls_disconnect_if_change() {
    let mut test = Smoketest::builder()
        .precompiled_module("rls-no-filter")
        .autopublish(false)
        .build();

    let name = format!("test-db-{}", std::process::id());

    // Initial publish without RLS
    test.publish_module_named(&name, false).unwrap();

    // Now re-publish with RLS added (requires --break-clients)
    test.use_precompiled_module("rls-with-filter");
    test.publish_module_with_options(&name, false, true).unwrap();

    // Check the row-level SQL filter is added correctly
    test.assert_sql(
        "SELECT sql FROM st_row_level_security",
        r#" sql
------------------------------------------------
 "SELECT * FROM users WHERE identity = :sender""#,
    );

    let logs = test.logs(100).unwrap();

    // Validate disconnect + schema migration logs
    assert!(
        logs.iter().any(|l| l.contains("Disconnecting all users")),
        "Expected 'Disconnecting all users' in logs: {:?}",
        logs
    );
}

/// Tests that not changing the RLS rules does not disconnect existing clients
#[test]
fn test_rls_no_disconnect() {
    let mut test = Smoketest::builder()
        .precompiled_module("rls-with-filter")
        .autopublish(false)
        .build();

    let name = format!("test-db-{}", std::process::id());

    // Initial publish with RLS
    test.publish_module_named(&name, false).unwrap();

    // Re-publish the same module (no RLS change)
    test.publish_module_named(&name, false).unwrap();

    let logs = test.logs(100).unwrap();

    // Validate no disconnect logs
    assert!(
        !logs.iter().any(|l| l.contains("Disconnecting all users")),
        "Expected no 'Disconnecting all users' in logs: {:?}",
        logs
    );
}
